### DISCUSSION TITLE
RTKQ & RTKM mocking and jest test case for covering?

### DISCUSSION THREAD

**Discussion by yaswanth1121997** [reduxjs/redux-toolkit#3287] [Q&A]
```
import { createSlice, createAction } from '@reduxjs/toolkit';
import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';

export const myApi = createApi({
    reducerPath: 'myApiPath',
    baseQuery: fetchBaseQuery({
        baseUrl: 'host url'
    }),
    endpoints: builder => ({
        getAllData: builder.query({
            query: () => 'endpoint url'
        })
    })
});

const mySlice = createSlice({
    name: 'mySliceName',
    initialState: {
        testState: {}
    },
    reducers: {
        // some reducers
    },
    extraReducers: builder => {
        builder.addMatcher(
            myApi.endpoints.getAllData.matchFulfilled,
            (state, { payload }) => {
                // doing something
            }
        );
    }
});

export default mySlice.reducer;
export const { useGetAllData } = myApi;
```
Can someone help me How to write the jest test case for covering this entire file especially extraReducers matchCases like fulfilled & RTKQ ?

**Comment by markerikson** [comment]
We specifically recommend writing tests that do mocking at the network request level:

https://redux.js.org/usage/writing-tests

**Reply by yaswanth1121997** [reply]
I tried that way, but didn't helped. Can you provide the working example for this ?

**Reply by phryneas** [reply]
If you provide what you tried, and which errors you encountered we can point out what you are doing wrong. Apart from that, you can look into our tests how we do it. But we are not writing a full example for you from zero here.

**Reply by ssk-17** [reply]
@phryneas Do we have any way to test the myApi.endpoints.getAllData.matchFulfilled in above case ? Thanks!
I have similar scenario:

`const optionsSlice = createSlice({
  name: 'options',
  initialState: INITIAL_STATE,
  reducers: {},
  extraReducers: builder => {
    builder.addMatcher(
      optionsApi.endpoints.getPlannerConfigOptions.matchFulfilled,
      (state, { payload }) => {
        console.log(
          'ssk17 getPlannerConfigOptions called, payload: ',
          JSON.stringify(payload),
        );
        state.countries = payload.country ?? INITIAL_STATE.countries;
        state.currencies = payload.currency ?? INITIAL_STATE.currencies;
        state.products = payload.product ?? INITIAL_STATE.products;
        state.maxPlanLength =
          payload.maxPlanLength ?? INITIAL_STATE.maxPlanLength;
        state.featureFlags = payload.featureFlags ?? INITIAL_STATE.featureFlags;
      },
    );
  },
});`

And this is my test file:

`describe('test getPlannerConfigOptions ', () => {
  let mockStore: any;
  beforeEach(() => {
    mockStore = configureStore({
        reducer: {
          options: optionsSlice.reducer,
          [optionsApi.reducerPath]: optionsApi.reducer,
        },
        middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(optionsApi.middleware),
      },
    );
  });

  it('test getPlannerConfigOptions', async () => {

    const mockResponse: XPROption = {
      country: [
        {
          'id': 2077456,
          'name': 'The Commonwealth of Australia',
          'translation': 'Australia',
          'countryCode': 'AU',
          'populationThreshold': 300000,
        },
        {
          'id': 3469034,
          'name': 'The Federative Republic of Brazil',
          'translation': 'Brazil',
          'countryCode': 'BR',
          'populationThreshold': 700000,
        },
        {
          'id': 6252001,
          'name': 'The United States of America',
          'translation': 'United States',
          'countryCode': 'US',
          'populationThreshold': 700000,
        },
      ],
      currency: [
        'BRL',
        'AUD',
        'USD',
      ],
      product: [
        {
          'value': 3,
          'label': 'Auction Reach',
          'maxDays': 93,
        },
        {
          'value': 5,
          'label': 'Auction Video views (6s)',
          'maxDays': 93,
        },
      ],
      'maxPlanLength': 225,
      featureFlags: {},
    };
    const expectedState = {
      countries: [
        {
          'id': 2077456,
          'name': 'The Commonwealth of Australia',
          'translation': 'Australia',
          'countryCode': 'AU',
          'populationThreshold': 300000,
        },
        {
          'id': 3469034,
          'name': 'The Federative Republic of Brazil',
          'translation': 'Brazil',
          'countryCode': 'BR',
          'populationThreshold': 700000,
        },
        {
          'id': 6252001,
          'name': 'The United States of America',
          'translation': 'United States',
          'countryCode': 'US',
          'populationThreshold': 700000,
        },
      ],
      currencies: [
        'BRL',
        'AUD',
        'USD',
      ],
      products: [
        {
          'value': 3,
          'label': 'Auction Reach',
          'maxDays': 93,
        },
        {
          'value': 5,
          'label': 'Auction Video views (6s)',
          'maxDays': 93,
        },
      ],
      pulseContextualTags: [],
      interestTags: [],
      locations: [],
      maxPlanLength: 225,
      featureFlags: {},
      population: {},
    };
    
    const action = {type: optionsApi.endpoints.getPlannerConfigOptions.matchFulfilled.name, payload: mockResponse};
    const  newState = optionsSlice.reducer(OPTIONS_INITIAL_STATE, action);

    return expect(newState).toEqual(expectedState);
  });
});`

But I don't see the expected state matching, I dont see the log statement that I added in my addMatcher too!

**Reply by markerikson** [reply]
@ssk-17 : per above, we specifically recommend using MSW to mock out the network request, then have the test just make the network request as usual (such as rendering the query hook in a component).  The `fulfilled` action will be dispatched, and that `matchFulfilled` case will run.

_Don't_ try to dispatch those actions manually - RTKQ fills out a lot of extra metadata that you're not including there.

**Comment by LuneLink** [comment]
@yaswanth1121997  i don't  know how much it help you but you can try do next 
```
    it('should ', () => {
      jest.spyOn(myApi.endpoints.getAllData.matchFulfilled, 'matchFulfilled').mockReturnValue(true);
      //todo: find the way to do action without mocking;
      const res = mySlice.reducer(initialState, { type: "any", payload: mockResponse });
      expect(res).toEqual(mockResponse);
    });
```

**Comment by vishalboine** [comment]
I'm also facing the same issue and I'm using msw which is giving me undefined response in my vitest environment. Do we need to set headers too while calling the handler for mock api

`http.get('API URL',  () => {
      return HttpResponse.json({
        "statusCode": 200,
        <<api mock data>>
    }, {status: 200})`

**Reply by phryneas** [reply]
Usually not

**Reply by aryaemami59** [reply]
@vishalboine Can you give me the versions of `msw` and `vitest` and a code snippet or maybe a small repro? I can take a look at this but I need more info because `msw`'s behavior has changed a lot between updates.
